torch/distributions/kl.py

```diff
@@ -172,14 +172,13 @@ def _kl_beta_beta(p, q):
 @register_kl(Dirichlet, Dirichlet)
 def _kl_dirichlet_dirichlet(p, q):
     # From http://bariskurt.com/kullback-leibler-divergence-between-two-dirichlet-and-beta-distributions/
-    sum_p_alpha = p.alpha.sum(-1)
-    sum_q_alpha = q.alpha.sum(-1)
-    t1 = torch.lgamma(sum_p_alpha)
-    t2 = torch.lgamma(sum_q_alpha)
-    t3 = p.alpha.lgamma().sum(-1)
-    t4 = q.alpha.lgamma().sum(-1)
-    t5 = (p.alpha - q.alpha) * (p.alpha.digamma() - sum_p_alpha.digamma())
-    return t1 - t3 - t2 + t4 + t5.sum(-1)
+    sum_p_alpha = p.alpha.sum(-1).expand_as(p.alpha)
+    sum_q_alpha = q.alpha.sum(-1).expand_as(q.alpha)
+    t1 = sum_p_alpha.lgamma() - p.alpha.lgamma().sum(-1)
+    t2 = sum_q_alpha.lgamma() - q.alpha.lgamma().sum(-1)
+    t3 = ((p.alpha - q.alpha) * p.alpha.digamma()).sum(-1)
+    t4 = sum_p_alpha * (p.alpha - q.alpha).sum(-1)
```

Comment:
Should the `sum_p_alpha` here be instead `sum_p_alpha.digamma().unsqueeze(-1)`?

Reply:
Oh, yes. Sorry about that.

Reply:
Is this the last remaining kl implementation that fails the tests? Have you tried it with very high sample count, say 1e7? It might be numerically stable to keep `(p.alpha.digamma - sum_p_alpha.digamma().unsqueeze(-1))` in a single term.

Reply:
For 1e6 the test fails for `KL(Dirichlet, Dirichlet)`.